src/robot/Robot.cpp

```diff
@@ -1,21 +1,454 @@
+    {
+      dtwarn << "CollisionFreePtr space does not match robot space."
+             << std::endl;
+      return selfCollisionFree;
+    }
+    constraints.emplace_back(collisionFree);
+  }
+
+  return std::make_shared<TestableIntersection>(mStateSpace, constraints);
+}
+
+// ==============================================================================
+constraint::TestablePtr Robot::getWorldCollisionConstraint(
+    const std::vector<std::string> bodyNames) const
+{
+  // Only use root robot's self-collision constraint and World
+  if (mRootRobot)
+  {
+    return mRootRobot->getWorldCollisionConstraint(bodyNames);
+  }
+
+  if (!mWorld)
+    return getSelfCollisionConstraint();
+
+  auto robotCollisionGroup
+      = mCollisionDetector->createCollisionGroup(mMetaSkeleton.get());
+
+  auto worldCollisionGroup = mCollisionDetector->createCollisionGroup();
+  for (std::string name : bodyNames)
+  {
+    auto skeleton = mWorld->getSkeleton(name);
+    if (skeleton)
+    {
+      worldCollisionGroup->addShapeFramesOf(skeleton.get());
+    }
+  }
+
+  // Add all skeletons in world if bodyNames is empty
+  if (bodyNames.size() == 0)
+  {
+    for (std::size_t i = 0; i < mWorld->getNumSkeletons(); i++)
+    {
+      auto skeleton = mWorld->getSkeleton(i);
+      if (skeleton != mMetaSkeleton->getBodyNode(0)->getSkeleton())
+      {
+        worldCollisionGroup->addShapeFramesOf(skeleton.get());
+      }
+    }
+  }
+
+  // Create constraint and combine with self-constraint
+  auto worldCollisionConstraint
+      = std::make_shared<aikido::constraint::dart::CollisionFree>(
+          mStateSpace, mMetaSkeleton, mCollisionDetector);
+  return combineCollisionConstraint(worldCollisionConstraint);
+}
+
+//=============================================================================
+std::shared_ptr<Robot> Robot::registerSubRobot(
+    const dart::dynamics::MetaSkeletonPtr& metaSkeleton,
+    const std::string& name)
 {
-  return std::const_pointer_cast<statespace::dart::MetaSkeletonStateSpace>(
-      const_cast<const Robot*>(this)->getStateSpace());
+  // Ensure name is unique
+  if (mSubRobots.find(name) != mSubRobots.end())
+  {
+    dtwarn << "Subrobot '" << name << "' already exists." << std::endl;
+    return nullptr;
+  }
+
+  // Ensure subrobot DoFs are disjoint
+  for (const auto& subrobot : mSubRobots)
+  {
+    auto dofs = subrobot.second->mDofs;
+    for (std::string dofName : util::dofNamesFromSkeleton(metaSkeleton))
+    {
+      if (dofs.find(dofName) != dofs.end())
+      {
+        dtwarn << "Subrobot '" << name << "'' overlaps existing subrobot "
+               << subrobot.first << "." << std::endl;
+        return nullptr;
+      }
+    }
+  }
+
+  // Create the subrobot.
+  auto subRobot = std::make_shared<Robot>(metaSkeleton, name);
+  const auto& root = mRootRobot ? mRootRobot : RobotPtr(this);
+  subRobot->setRootRobot(root);
+  subRobot->setWorld(nullptr); // inherit root robot world
```

Comment:
This comment was a bit confusing to be because it seems to imply that because the world ptr is nullptr, the subrobot will inherit the root robot's world. But I couldn't find logic for that. Looking back in the code, I saw on line 202 that the subrobot *would* inherit the root robot's world, but that is not because of the nullptr, but rather because the world is set. So if you can change this comment to something like "The subrobot will inherit the root robot's world, hence its world is unnecessary" that might make it clearer